### PR TITLE
Integrate service-experiments component

### PR DESCRIPTION
**Pull request by travis79**
This is the implementation of Bugzilla [Bug 1543985](https://bugzilla.mozilla.org/show_bug.cgi?id=1543985)

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/fenix/blob/master/CHANGELOG.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt [hunk 1]
```diff
@@ -287,6 +289,13 @@ class GleanMetricsService(private val context: Context) : MetricsService {
             }
 
             activationPing.checkAndSend()
+
+            // We need to ensure this function is called after Glean is initialized because the service-experiments
+            // component requires Glean to be initialized first in case it needs to record metrics.
+            Experiments.initialize(
```
Comment (Dexterp37) [inline app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt]: Please note that `GleanMetricsService.start()` is being called every single time the telemetry option is toggled on/off in the UI.  After [this part](https://github.com/mozilla-mobile/fenix/blob/390659fe58fa99064b051983f39d183636a8e397/app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt#L264) it only gets called once. Is there any mechanism to disable experiments when telemetry is turned off? Is this needed? Does it need to be hooked up to a separate pref? That's probably a question for @mythmon (maybe we want a similar behaviour as Desktop?) and for Fenix people :)

**Comment by travis79** [comment]
> I'd like to see the comment below explicitly addressed: do we need to turn experiments on/off when turning telemetry off? Do we need a separate pref? The integration itself looks ok, but I'd really talk to a Fenix engineer about that too.

Yeah, that was my plan was to try and get someone like @boek to see what he thought about the best way to go about this after I ran into this yesterda.  

There really wasn't another good place to ensure that Glean was initialized since there currently isn't a callback mechanism built into the metrics service wrapper.  Also, it appears that Fenix is loading fretboard without a way to enable/disable it with a pref, so I just followed along since this should eventually replace fretboard in Fenix.

File: app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt (hunk 1 above)
Comment (mythmon) [reply app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt]: On Desktop, Normandy (which includes an experiments subsystem, but also other things like gradual rollouts) currently disables itself entirely when Telemetry is disabled. This has proven to be a problem for a few reasons.

The first is that there are many cases, where we might want the experiments program to be active without telemetry, such as QA. For example, mozregression builds have Telemetry disabled, and so Normandy is difficult to run regressions on.

The second is because Normandy does more than just experiments, there are plenty of times where we want to deliver something to users even if they don't have Telemetry enabled (Armagaddon, for a recent example).

The second doesn't apply to this integration, but the first does. I'd suggest making it independent of Telemetry. For a settings UI it might make sense to tie the two preferences together into one toggle, since I imagine that's what most users want.

Comment (travis79) [reply app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt]: Totally agree with keeping this separate from the telemetry preference based on that logic.  Do we want to have a completely separate preference here or treat it as always on like fretboard currently is treated in Fenix?

Comment (mythmon) [reply app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt]: I know that there are users that are ok keeping Telemetry on, but don't want experiments running. They are happy to contribute technical data about the product, but don't want it changing in unexpected ways. I think offering a toggle, even if it is in some "advanced" panel, would be a nice thing to offer to users.

I also don't really know how this affects the implementation. I'm still getting my bearings. If this is really hard then I think the fretboard approach would be a fine compromise.

Comment (travis79) [reply app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt]: This is the first time I've had to touch the Fenix codebase, so it's hard to say what the effort of adding a new pref/toggle will be but I'm looking into it before I decide which way to go.

Comment (Dexterp37) [reply app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt]: @travis79 this is a product specific question, so I'd recommend to also bring in this discuss a representative from the Fenix team.

> The first is that there are many cases, where we might want the experiments program to be active without telemetry, such as QA. For example, mozregression builds have Telemetry disabled, and so Normandy is difficult to run regressions on.

Does this apply here, though? When performing validation tasks, recently, QA used released binaries or nightly builds (both of them having telemetry on). I'm not sure there's a case in which one would have telemetry off outside of local testing builds.

Comment (travis79) [reply app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt]: I don't guess it had dawned on me until just now that this may be a moot point about having the init behind it's own preference because currently, Experiments requires Glean to be initialized first or it just won't init itself, see [here](https://github.com/mozilla-mobile/android-components/blob/663e67fc17bf3747afb76f202dc05627501cc2a3/components/service/experiments/src/main/java/mozilla/components/service/experiments/Experiments.kt#L65).

I'm not sure we can maintain a hard dependency on Glean like that if we plan on allowing the two libraries to be used separately from each other.

Comment (travis79) [reply app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt]: This may also benefit from the changes to recording metrics prior to initialization that are being proposed for the crash counts [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1566917) and may make the initialization order a little easier to handle.

**Comment by travis79** [comment]
This will now require [this](https://github.com/mozilla-mobile/android-components/pull/3821) to land and be available from android-components before this can be merged.

File: app/src/main/java/org/mozilla/fenix/FenixApplication.kt [hunk 2]
```diff
@@ -65,6 +66,12 @@ open class FenixApplication : Application() {
         // on the main process, as it uses Gecko to fetch experiments from the server.
         experimentLoader = loadExperiments()
 
+        // Enable the service-experiments component
+        Experiments.initialize(
+            applicationContext,
+            mozilla.components.service.experiments.Configuration(httpClient = lazy { components.core.client })
```
Comment (csadilek) [inline app/src/main/java/org/mozilla/fenix/FenixApplication.kt]: Just a heads-up that we have a potential problem here: `components.core.client` is using the Gecko fetch client and could be the first thing that needs a Gecko runtime. The runtime must be initialized on the main thread and this change could potentially cause initialization of the lazy `httpClient` (and therefore the Gecko runtime) on a background thread. This would happen if `service-experiments` is the first thing needing a Gecko runtime (which is likely as the Home fragment doesn't need one).

In `service-fretboard` the field isn't lazy and the httpClient is therefore initialized on the main thread: https://github.com/mozilla-mobile/fenix/blob/master/app/src/main/java/org/mozilla/fenix/FenixApplication.kt#L139

TL;DR: I think we need a change in Fenix to make sure the Gecko runtime is only ever created on the main thread before this change lands.
//cc @boek @pocmo @hawkinsw (as we've talked about this before)

Comment (pocmo) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt]: 👍 https://github.com/mozilla-mobile/android-components/issues/3741

Comment (travis79) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt]: @csadilek thanks for the heads up!  So would it be safer to just remove the lazy instantiation here or do I need to wait until this issue is resolved?

Comment (travis79) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt]: Also, I have other reasons to wait on this PR until https://github.com/mozilla-mobile/android-components/pull/3821 lands in A-C and is available in Fenix.

Comment (csadilek) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt]: @travis79 I think this is something we have to fix/guarantee in Fenix either way. However, there's another race condition in the `service-fretboard` integration and I was curious how that's best addressed using `service-experiments`. Experiments are loaded from storage and fetched from the network on startup (both on IO threads), but other components / parts of the app will check `isInExperiment` on the main thread (this currently happens on startup as well). There's no guarantee that `isInExperiment` will be called after the experiments have been read from storage: https://github.com/mozilla-mobile/fenix/blob/5ecdce6ba6c121cf15d915c12b52803cf564be95/app/src/main/java/org/mozilla/fenix/components/BackgroundServices.kt#L73

The way I see it, we need a mechanism to either block on experiments being loaded from storage (not network) or some other callback solution? @hawkinsw provided a solution that isn't currently used though: https://github.com/mozilla-mobile/fenix/blob/5ecdce6ba6c121cf15d915c12b52803cf564be95/app/src/main/java/org/mozilla/fenix/FenixApplication.kt#L97 

We might be using this incorrectly though. Wdyt?

Comment (travis79) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt]: I see what you mean, you don't really want to block in `onCreate()` so it's calling the `loadExperiments()` function of Fretboard on the IO thread.  

I think service-experiments may suffer from the same problem as it does require a little startup to read from storage, the difference being that service-experiments only stores the active experiment rather than the entire `experimentsResult` as retrieved from the server, and it stores it in SharedPreferences rather than a flat file.  The other difference is that service-experiments is using WorkManager to do the background work of fetching from the server so it shouldn't be slowing down the main thread. Once the WorkManager Worker receives the response from the server, it updates the active experiment info, but it doesn't currently have a mechanism to notify the application.

Being able to load this asynchronously is definitely going to lower startup impact and that could require some way to synchronize on loading the experiments before we go checking to see if we are in an experiment/branch.  I'm not sure if it's too much to ask of the experiments library or fretboard to handle this, or if it's more flexible to deal with it in the app like @hawkinsw's solution.

Personally, I like @hawkinsw's solution as it gives the app more control over how this will be executed in the background and awaited.

**Comment by travis79** [comment]
> This looks good to me but I think this is really blocked on [this comment](https://github.com/mozilla-mobile/fenix/pull/4098/files#r306065907).

I think that's been resolved now: https://github.com/mozilla-mobile/android-components/issues/3741

File: app/src/main/java/org/mozilla/fenix/FenixApplication.kt (hunk 2 above)
Comment (travis79) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt]: I filed [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1569647) to follow up on the callback question.

File: app/src/main/java/org/mozilla/fenix/FenixApplication.kt [hunk 3]
```diff
@@ -66,6 +67,12 @@ open class FenixApplication : Application() {
...
```
Comment (colintheshots) [inline app/src/main/java/org/mozilla/fenix/FenixApplication.kt]: If this will always be initialized on the main thread in service-experiments, then there's no need for the default lazy delegate thread safety. You could use `lazy(LazyThreadSafetyMode.NONE)`.

Comment (travis79) [reply app/src/main/java/org/mozilla/fenix/FenixApplication.kt]: Sounds reasonable to me.  Modification to remove the lazy delegate thread safety has been made.

**Comment by travis79** [comment]
Hi @colintheshots I was wondering if this might have slipped through the cracks. I made the change you requested and I was hoping you could make another review so that I can close this soon.  Thanks!

**Comment by colintheshots** [comment]
Thanks for making the change!
